File: client-android/serenada-core/src/main/java/app/serenada/core/call/SessionClock.kt
package app.serenada.core.call

internal interface SessionClock {
    /** Wall-clock milliseconds since epoch. */
    fun nowMs(): Long

    /**
     * Monotonic milliseconds for interval math: unaffected by
     * a wall-clock / NTP step, so a backward correction during an open dropout
     * can't record a real outage as 0ms. Production uses
     * `SystemClock.elapsedRealtime()`.
     */
    fun monotonicMs(): Long
}

File: client-ios/SerenadaCore/Sources/Call/SessionClock.swift
import Combine
import Foundation

@MainActor
protocol SessionClock: AnyObject {
    /// Wall-clock milliseconds since epoch.
    func nowMs() -> Int64
    /// Monotonic milliseconds for interval math: unaffected
    /// by a wall-clock / NTP step, so a backward correction during an open
    /// dropout can't record a real outage as 0ms. Production uses
    /// `DispatchTime` (CLOCK_UPTIME / mach absolute time).
    func monotonicMs() -> Int64
    func sleep(nanoseconds: UInt64) async throws
    func scheduleRepeating(intervalSeconds: TimeInterval, action: @escaping @MainActor () -> Void) -> AnyCancellable
}
